File: Sherlock-and-the-Valid-String.cpp
#include <bits/stdc++.h>

using namespace std;

/*
 * Complete the 'isValid' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */

string isValid_ByMap(string s)
{
    if (s.size() == 1) return "YES";

    unordered_map<char, int> charCounts;
    for (const auto& c : s) {
        charCounts[c]++;
    }

    int minCount = INT_MAX;
    int maxCount = INT_MIN;
    for (const auto& entry : charCounts) {
        minCount = min(minCount, entry.second);
        maxCount = max(maxCount, entry.second);
    }

    if (minCount == maxCount) return "YES";

    int minCountFreq = 0;
    int maxCountFreq = 0;
    for (const auto& entry : charCounts) {
        if (entry.second == minCount) minCountFreq++;
        if (entry.second == maxCount) maxCountFreq++;
    }

    if ((maxCount - minCount == 1 && maxCountFreq == 1) ||
        (minCount == 1 && minCountFreq == 1 && maxCountFreq == charCounts.size() - 1))
        return "YES";

    return "NO";
}


const int AlphabetLength = 'z' - 'a' + 1;

string isValid_ByVector(string s)
{
    if (s.size() == 1) return "YES";

    vector<int> charCounts(AlphabetLength);
    for (char c : s)
    {
        charCounts[c - 'a']++;
    }

    sort(charCounts.begin(), charCounts.end());

    int firstNonzero = 0;
    while (firstNonzero < AlphabetLength && charCounts[firstNonzero] == 0)
    {
        firstNonzero++;
    }
    
    if (firstNonzero >= AlphabetLength) return "NO";

    int minCount = charCounts[firstNonzero];
    int maxCount = charCounts[AlphabetLength - 1];

    if (minCount == maxCount) return "YES";

    if ((maxCount - minCount == 1 && maxCount > charCounts[AlphabetLength - 2]) ||
        (minCount == 1 && charCounts[firstNonzero + 1] == maxCount))
        return "YES";

    return "NO";
}


string isValid(string s)
{
    if( s.size() > 100)
    {
        return isValid_ByMap(s);
    }
    else
    {
        return isValid_ByVector(s);
    }
}


int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string s;
    getline(cin, s);

    string result = isValid(s);

    fout << result << "\n";

    fout.close();

    return 0;
}

File: Sherlock-and-the-Valid-String.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

class Result {

    /*
     * Complete the 'isValid' function below.
     *
     * The function is expected to return a STRING.
     * The function accepts STRING s as parameter.
     */

    private static String isValidByMap(String s) {
        if (s.length() == 1) return "YES";

        Map<Character, Integer> charCounts = new HashMap<>();
        for (char c : s.toCharArray()) {
            charCounts.put(c, charCounts.getOrDefault(c, 0) + 1);
        }

        int minCount = Integer.MAX_VALUE;
        int maxCount = Integer.MIN_VALUE;
        for (int count : charCounts.values()) {
            minCount = Math.min(minCount, count);
            maxCount = Math.max(maxCount, count);
        }

        if (minCount == maxCount) return "YES";

        int minCountFreq = 0;
        int maxCountFreq = 0;
        for (int count : charCounts.values()) {
            if (count == minCount) minCountFreq++;
            if (count == maxCount) maxCountFreq++;
        }

        if ((maxCount - minCount == 1 && maxCountFreq == 1) ||
                (minCount == 1 && minCountFreq == 1 && maxCountFreq == charCounts.size() - 1))
            return "YES";

        return "NO";
    }
    
    
    private static final int AlphabetLength = 'z' - 'a' + 1;

    private static String isValidByVector(String s) {
        if (s.length() == 1) return "YES";

        int[] charCounts = new int[AlphabetLength];
        for (char c : s.toCharArray()) {
            charCounts[c - 'a']++;
        }

        Arrays.sort(charCounts);

        int firstNonzero = 0;
        while (firstNonzero < AlphabetLength && charCounts[firstNonzero] == 0) {
            firstNonzero++;
        }

        if (firstNonzero >= AlphabetLength) return "NO";

        int minCount = charCounts[firstNonzero];
        int maxCount = charCounts[AlphabetLength - 1];

        if (minCount == maxCount) return "YES";

        if ((maxCount - minCount == 1 && maxCount > charCounts[AlphabetLength - 2]) ||
                (minCount == 1 && charCounts[firstNonzero + 1] == maxCount))
            return "YES";

        return "NO";
    }

    public static String isValid(String s) {
        if (s.length() > 100) {
            return isValidByMap(s);
        } else {
            return isValidByVector(s);
        }
    }
}

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String s = bufferedReader.readLine();

        String result = Result.isValid(s);

        bufferedWriter.write(result);
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}
